tony-history-server/app/utils/ParserUtils.java

```diff
@@ -30,28 +32,52 @@
 public class ParserUtils {
   private static final Logger.ALogger LOG = Logger.of(ParserUtils.class);
 
-  public static JobMetadata parseMetadata(FileSystem fs, Path metadataFilePath) {
-    if (!pathExists(fs, metadataFilePath)) {
-      return new JobMetadata();
+  @VisibleForTesting
+  static boolean isValidMetadata(String[] metadata, String jobIdRegex) {
+    if (metadata.length != 5) {
+      LOG.error("Missing fields in metadata");
+      return false;
     }
+    return metadata[0].matches(jobIdRegex) && metadata[1].matches("\\d+") && metadata[2].matches("\\d+")
+        && metadata[3].matches(metadata[3].toLowerCase()) && metadata[4].equals(metadata[4].toUpperCase());
+  }
 
-    String fileContent = contentOfHdfsFile(fs, metadataFilePath);
+  public static JobMetadata parseMetadata(FileSystem fs, Path jobFolderPath, String jobIdRegex) {
     JobMetadata jobMetadata = new JobMetadata();
-    JsonNode jObj;
+    DateFormat simple = new SimpleDateFormat("dd MMM yyyy HH:mm:ss:SSS Z");
+    String[] metadata;
+
+    if (!pathExists(fs, jobFolderPath)) {
+      return jobMetadata;
+    }
+
     try {
-      jObj = Json.parse(fileContent);
-    } catch (JsonSyntaxException e) {
-      LOG.error("Couldn't parse metadata", e);
+      metadata = Arrays.stream(fs.listStatus(jobFolderPath))
+          .filter(f -> f.getPath().toString().contains("jhist")) // ignore config.xml
+          .map(histFilePath -> HdfsUtils.getJobId(histFilePath.toString()))
+          .map(histFile -> histFile.substring(0, histFile.lastIndexOf('.'))) // remove .jhist
+          .map(histFileNoExt -> histFileNoExt.split("-"))
+          .flatMap(Arrays::stream)
+          .toArray(String[]::new);
+    } catch (IOException e) {
+      LOG.error("Failed to scan " + jobFolderPath.toString(), e);
       return jobMetadata;
     }
-    jobMetadata.setId(jObj.get("id").textValue());
-    jobMetadata.setJobLink(jObj.get("url").textValue());
-    jobMetadata.setConfigLink("/jobs/" + jobMetadata.getId());
-    jobMetadata.setStarted(jObj.get("started").textValue());
-    jobMetadata.setCompleted(jObj.get("completed").textValue());
-    jobMetadata.setCompleted(jObj.get("completed").textValue());
-    jobMetadata.setStatus(jObj.get("status").textValue());
-    jobMetadata.setUser(jObj.get("user").textValue());
+
+    if (!isValidMetadata(metadata, jobIdRegex)) {
+      // this should never happen unless user rename the history file
+      LOG.error("Metadata isn't valid");
+      return jobMetadata;
+    }
+
+    jobMetadata.setId(metadata[0]);
+    jobMetadata.setJobLink("/jobs/" + jobMetadata.getId());
+    jobMetadata.setConfigLink("/config/" + jobMetadata.getId());
+    jobMetadata.setStarted(simple.format(new Date(Long.parseLong(metadata[1]))));
+    jobMetadata.setCompleted(simple.format(new Date(Long.parseLong(metadata[2]))));
```

Comment:
I think JobMetadata should store `long` epoch timestamps rather than Strings. At UI rendering time, we can decide what format / timezone to show the time in. I think this is converting the timestamps to a final output representation too early.

Reply:
See https://github.com/linkedin/TonY/pull/81/commits/6ff6714324def1e385f5995d76f389a48c1eaf15 for the updated implementation of `JobMetadata`. With this being the new implementation, do you think this comment still apply?

Reply:
That commit still seems to be storing times as Strings rather than longs. I think we should store times as longs. It's a more machine-readable representation. Do the conversion to a human-readable representation later depending on what timezone/datetime format we want to render.

Reply:
So right now the UI call getters to retrieve the time. So if we were to store times as longs, we would have to do the date parsing in getters, which looks unusual imo (https://gist.github.com/pdtran3k6/d19e233cf876e8115a7a2db82c7ac32e), but we can certainly do that if that's more maintainable for you guys.

Reply:
MR's JobInfo stores times internally as longs. It then provides getXXXTimeStr() methods (in addition to `long getXXXTime()` methods) that return as Strings (in some predefined format). I think we can do something similar.

Reply:
Is it standard to store these as `Date` objects?

Reply:
Or I can just convert + format it when UI fetches `started`/`completed`. Also removed some unused methods: https://github.com/linkedin/TonY/pull/81/commits/d150f40c5181d7424751119a86fc9954da1cfd17.

Reply:
Nice, that's better. `Date` is a good choice.